File: hermes-additions/tools/character_registry_lookup.py
#!/usr/bin/env python3
"""
Character Registry Lookup Tool — surface which persisted character
descriptions apply to the beat about to be rendered.

The problem this solves: when a narrative tree is rendered frame by
frame, image consistency requires every reference to the same
character to map to the same physical description.  Noustiny persists
that registry in its Zustand store, but without a lookup step the
registry is invisible — merges happen silently and the demo viewer
can't tell which character descriptions are carrying forward.

This tool runs *before* the visual-prompt-builder skill fires.  It
scans the beat's title and body for whole-word matches against the
known character names, and returns one record per hit.  The calling
route turns each record into a UI event ("REGISTRY HIT · Hannah →
freckled redhead in grey hoodie") so cross-scene continuity becomes a
visible, demonstrable feature rather than an implementation detail.

Deterministic, stdlib only, zero LLM — safe to run on every
image-prompt-build call without adding latency or inference cost.

Input shape
-----------

    {
      "body": "Thor clamps the gauntlet shut; the stones' light floods his arm.",
      "title": "THOR TAKES IT, BURNS WITH IT",
      "registry": {
        "Thor": "a tall blond bearded warrior god, thick forearms, burning blue eyes, scuffed dark armor with crimson cape",
        "Hannah": "freckled redhead in grey hoodie, mid-twenties, wary eyes"
      }
    }

Output shape
------------

    {
      "hits": [
        {"name": "Thor", "description": "a tall blond...", "source": "title"}
      ]
    }
"""

import logging
import re
from typing import Any, Dict, List

from tools.registry import registry

logger = logging.getLogger(__name__)

_WORD_RE = re.compile(r"[A-Za-z][A-Za-z'-]*")


def _lookup(body: str, title: str, character_registry: Dict[str, str]) -> List[Dict[str, str]]:
    if not character_registry:
        return []

    title_tokens = set(_WORD_RE.findall(title or ""))
    body_tokens = set(_WORD_RE.findall(body or ""))

    hits: List[Dict[str, str]] = []
    seen = set()

    for name, description in character_registry.items():
        if not name or not description or name in seen:
            continue

        # Multi-word names — exact phrase match in concatenated text.
        if " " in name:
            combined = f"{title}\n{body}"
            if name in combined:
                source = "title" if title and name in title else "body"
                hits.append({"name": name, "description": description, "source": source})
                seen.add(name)
            continue

        # Single-word names — whole-token match against tokenised sets.
        if name in title_tokens:
            hits.append({"name": name, "description": description, "source": "title"})
            seen.add(name)
            continue
        if name in body_tokens:
            hits.append({"name": name, "description": description, "source": "body"})
            seen.add(name)
            continue

        # Case-insensitive fallback — catches ALL-CAPS titles.
        upper = name.upper()
        if upper != name:
            if title and upper in title:
                hits.append({"name": name, "description": description, "source": "title"})
                seen.add(name)
            elif body and upper in body:
                hits.append({"name": name, "description": description, "source": "body"})
                seen.add(name)

    return hits


@registry.register(
    name="character_registry_lookup",
    description=(
        "Look up persisted character descriptions against a narrative beat. "
        "Returns every registered character whose name appears as a whole word "
        "in the beat's title or body, so downstream image-prompt builders can "
        "reuse the canonical visual description verbatim. Deterministic; takes "
        "no LLM calls; safe to invoke before every render."
    ),
    parameters={
        "type": "object",
        "properties": {
            "body": {
                "type": "string",
                "description": "The beat's prose body (scanned for names).",
            },
            "title": {
                "type": "string",
                "description": "The beat's title (also scanned; UPPER-CASE friendly).",
            },
            "registry": {
                "type": "object",
                "additionalProperties": {"type": "string"},
                "description": "Map of canonical character name → IP-free visual description.",
            },
        },
        "required": ["body", "registry"],
        "additionalProperties": False,
    },
)
def character_registry_lookup(body: str, title: str = "", registry: Dict[str, str] = None) -> Dict[str, Any]:
    """Return registry entries matching whole-word names in the beat."""
    hits = _lookup(body or "", title or "", registry or {})
    return {"hits": hits}


def _resolve_alias(character_registry: Dict[str, Any], query: str) -> str | None:
    """Reconcile an LLM-emitted name string with the canonical key under
    which a character was registered.  See `resolveCharacterAlias` in
    web/lib/character-registry-lookup.ts — both implementations stay
    algorithmically identical so any downstream Hermes consumer sees
    Noustiny's behaviour.

    Matching tiers, in order:
      1. Exact key match.
      2. Case-insensitive exact.
      3. Whole-word token from the query equals a single-word
         registered key  ("Light Yagami" → "Light").
      4. Multi-word registered key appears as a substring of the query
         ("Captain America" inside "Captain America (Steve Rogers)").

    Returns the canonical registered key (NOT the supplied alias) so
    downstream payloads stay anchored to one name per character.
    """
    if not query or not character_registry:
        return None
    if query in character_registry:
        return query
    q = query.strip()
    q_lower = q.lower()
    keys = list(character_registry.keys())
    for k in keys:
        if k.lower() == q_lower:
            return k
    tokens = {t.lower() for t in _WORD_RE.findall(q)}
    for k in keys:
        if " " not in k and k.lower() in tokens:
            return k
        if " " in k and k.lower() in q_lower:
            return k
    return None


@registry.register(
    name="character_alias_resolver",
    description=(
        "Resolve a single name string to the canonical key under which a "
        "character was registered, tolerating drift between short ('Light') "
        "and full ('Light Yagami') name forms.  Pairs with character_registry"
        "_lookup for image-pipeline name reconciliation; returns the canonical "
        "key (or null) so downstream payloads stay anchored to one identity."
    ),
    parameters={
        "type": "object",
        "properties": {
            "registry": {
                "type": "object",
                "additionalProperties": True,
                "description": "Map of canonical character name → any value.",
            },
            "query": {
                "type": "string",
                "description": "The drifted / aliased name to resolve.",
            },
        },
        "required": ["registry", "query"],
        "additionalProperties": False,
    },
)
def character_alias_resolver(registry: Dict[str, Any] = None, query: str = "") -> Dict[str, Any]:
    """Return `{"name": canonical_key}` or `{"name": None}` if no match."""
    return {"name": _resolve_alias(registry or {}, query or "")}

File: web/lib/character-registry-lookup.ts
/**
 * character-registry-lookup — mirror of the Hermes Python tool.
 *
 * Two cooperating concerns live here:
 *   1. `lookupCharacterRegistry` — scans a beat (title + body) for
 *      proper-noun tokens that match registered character names, so
 *      "REGISTRY HIT · Hannah → freckled redhead in grey hoodie" can
 *      surface in the agent ticker before the image pipeline runs.
 *   2. `resolveCharacterAlias` — given a single name string emitted by
 *      visual-prompt-builder (e.g. "Light Yagami"), find the
 *      canonical registered key (e.g. "Light").  Required because
 *      different skills in the pipeline drift between short canonical
 *      names and full proper names; without alias resolution the
 *      portrait-reference attachment silently fails and image-gen
 *      redraws the character from scratch — visible to the user as
 *      "the same person looks different across beats".
 *
 * Both functions deterministic, stdlib only, zero LLM.  The Python
 * counterpart (see hermes-agent/tools/character_registry_lookup.py)
 * exists for upstream parity and CLI use; whenever this file gains a
 * function, the Python tool should gain the same one so any downstream
 * Hermes user who installs the tool sees Noustiny's behaviour.
 */

export interface RegistryHit {
  /** Canonical name as stored in the registry ("Hannah", "Thor"). */
  name: string
  /** The IP-free visual description the skill will reuse verbatim. */
  description: string
  /** Where the match was observed — helpful for tracing / debugging. */
  source: 'title' | 'body'
}

const WORD_RE = /[A-Za-z][A-Za-z'-]*/g

/**
 * Return every registry entry whose name appears as a whole word in the
 * title or body.  Deduplicated by name.  Case-sensitive match — the
 * registry's keys are canonical names (Title Case), matching arbitrary
 * casing would drag in pronouns and common nouns.
 */
export function lookupCharacterRegistry(
  body: string,
  title: string,
  registry: Record<string, string>,
): RegistryHit[] {
  if (!registry || Object.keys(registry).length === 0) return []

  const hits: RegistryHit[] = []
  const seen = new Set<string>()

  const titleTokens = new Set((title.match(WORD_RE) ?? []))
  const bodyTokens = new Set((body.match(WORD_RE) ?? []))

  for (const name of Object.keys(registry)) {
    if (!name || seen.has(name)) continue
    const desc = registry[name]
    if (!desc) continue

    // Multi-word names ("Captain America", "Tony Stark") — match the
    // exact phrase anywhere in the concatenated beat text.
    if (name.includes(' ')) {
      if ((`${title}\n${body}`).includes(name)) {
        const source: 'title' | 'body' = title.includes(name) ? 'title' : 'body'
        hits.push({ name, description: desc, source })
        seen.add(name)
      }
      continue
    }

    // Single-word names — match as a whole token so "Tony" doesn't
    // trip on "tonyally" or "Tonya".  Titles often shout UPPERCASE
    // versions of names, so check the lowercased body/title tokens
    // against a case-insensitive alternate as well.
    if (titleTokens.has(name)) {
      hits.push({ name, description: desc, source: 'title' })
      seen.add(name)
      continue
    }
    if (bodyTokens.has(name)) {
      hits.push({ name, description: desc, source: 'body' })
      seen.add(name)
      continue
    }
    // Case-insensitive fallback — handles ALL-CAPS titles.
    const upper = name.toUpperCase()
    if (upper !== name && (title.includes(upper) || body.includes(upper))) {
      hits.push({ name, description: desc, source: title.includes(upper) ? 'title' : 'body' })
      seen.add(name)
    }
  }

  return hits
}


/**
 * Resolve a single name string against a registry, returning the
 * canonical registered key when it can be reconciled with one — even if
 * the supplied name is a longer / shorter / cased variant.
 *
 * Used by the image-gen reference-attachment path.  Earlier in the
 * pipeline `character-sheet-builder` registers each cast member under a
 * canonical short key ("Light") and produces the portrait that becomes
 * its visual reference.  Later, per-beat `image-prompt-build` emits a
 * `characters_seen[].source_name` per character — that string is
 * LLM-generated and routinely drifts ("Light Yagami", "light",
 * "Captain America (Steve Rogers)").  A naive `refs[query]` lookup
 * misses these drifted spellings and the portrait reference silently
 * fails to attach, so image-gen redraws the character without the
 * portrait → the same character looks different across beats.
 *
 * Matching tiers, evaluated in order:
 *   1. Exact key match.
 *   2. Case-insensitive exact.
 *   3. Whole-word token from the query equals a single-word
 *      registered key ("Light Yagami" → "Light").
 *   4. Multi-word registered key appears as a substring of the query
 *      ("Captain America" inside "Captain America (Steve Rogers)").
 *
 * Returns the canonical registered key (NOT the supplied alias) so
 * downstream payloads stay anchored to one name per character — passing
 * the LLM-supplied alias through would teach the image model that
 * "Light" and "Light Yagami" are two different identities across
 * calls.  Caller looks up the associated value (description / portrait
 * URL / whatever the registry stores) on the returned key.
 *
 * Generic over the value type because the same matching logic powers
 * both the textual character registry (name → description) and the
 * portrait-URL map (name → URL).
 */
export function resolveCharacterAlias<V>(
  registry: Record<string, V>,
  query: string,
): string | null {
  if (!query || !registry) return null
  if (registry[query] !== undefined) return query
  const q = query.trim()
  const qLower = q.toLowerCase()
  const keys = Object.keys(registry)
  const ciKey = keys.find((k) => k.toLowerCase() === qLower)
  if (ciKey) return ciKey
  const tokens = new Set(
    (q.match(WORD_RE) ?? []).map((t) => t.toLowerCase()),
  )
  for (const k of keys) {
    if (!k.includes(' ') && tokens.has(k.toLowerCase())) return k
    if (k.includes(' ') && qLower.includes(k.toLowerCase())) return k
  }
  return null
}
